kpi/views.py

```diff
@@ -748,15 +749,24 @@ def create(self, request, *args, **kwargs):
             asset_uid = self.get_parents_query_dict().get("asset")
             asset = get_object_or_404(self.parent_model, uid=asset_uid)
             instance_id = request.data.get("instance_id")
-            if not HookUtils.call_services(asset, instance_id):
+            instance = asset.deployment.get_submission(instance_id)
+
+            # Check if instance really belongs to Asset.
+            if not (instance and instance.get(asset.deployment.INSTANCE_ID_FIELDNAME) == instance_id):
+                response_status_code = status.HTTP_404_NOT_FOUND
+                response = {
+                    "detail": _("Resource not found")
+                }
+
+            elif not HookUtils.call_services(asset, instance_id):
                 response_status_code = status.HTTP_409_CONFLICT
                 response = {
                     "detail": _(
                         "Your data for instance {} has been already submitted.".format(instance_id))
```

Comment:
I was surprised to see a `409` returned if `call_services()` fails, but I realize now that `call_services()` doesn't (directly) call any services. It kicks off a Celery task and doesn't wait around to see what the result will be; it uses its return value to indicate whether a `HookLog` for `instance_id` already exists. Maybe the function could be renamed to better describe its workings